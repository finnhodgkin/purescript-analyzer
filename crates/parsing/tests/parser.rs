macro_rules! parser_tests {
    ($($name:ident => $source:expr),+ $(,)?) => {
        $(
            #[test]
            fn $name() {
                let lexed = lexing::lex($source);
                let tokens = lexing::layout(&lexed);
                let (node, errors) = parsing::parse(&lexed, &tokens);
                insta::assert_debug_snapshot!((node, errors));
            }
        )+
    };
}

parser_tests!(
    comment_pre_post => include_str!("parser/CommentPrePost.purs"),
    lexer_error_inserted => include_str!("parser/LexerErrorInserted.purs"),
    module_export => include_str!("parser/ModuleExport.purs"),
    module_export_class => include_str!("parser/ModuleExportClass.purs"),
    module_export_class_error_end => include_str!("parser/ModuleExportClassErrorEnd.purs"),
    module_export_comma_error => include_str!("parser/ModuleExportCommaError.purs"),
    module_export_data => include_str!("parser/ModuleExportData.purs"),
    module_export_data_comma_error => include_str!("parser/ModuleExportDataCommaError.purs"),
    module_export_data_item_after_double_period => include_str!("parser/ModuleExportDataItemAfterDoublePeriod.purs"),
    module_export_data_item_before_double_period => include_str!("parser/ModuleExportDataItemBeforeDoublePeriod.purs"),
    module_export_data_item_invalid => include_str!("parser/ModuleExportDataItemInvalid.purs"),
    module_export_empty => include_str!("parser/ModuleExportEmpty.purs"),
    module_export_invalid => include_str!("parser/ModuleExportInvalid.purs"),
    module_export_type_operator => include_str!("parser/ModuleExportTypeOperator.purs"),
    module_export_type_operator_error => include_str!("parser/ModuleExportTypeOperatorError.purs"),
    module_export_type_operator_error_end => include_str!("parser/ModuleExportTypeOperatorErrorEnd.purs"),
    module_header => include_str!("parser/ModuleHeader.purs"),
    module_header_prefixed => include_str!("parser/ModuleHeaderPrefixed.purs"),
);
