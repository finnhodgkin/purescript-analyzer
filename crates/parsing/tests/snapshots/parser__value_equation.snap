---
source: crates/parsing/tests/parser.rs
expression: "(node, errors)"
snapshot_kind: text
---
(
    Module@0..78
      ModuleHeader@0..28
        MODULE@0..6 "module"
        Comment@6..7
          WHITESPACE@6..7 " "
        ModuleName@7..21
          UPPER@7..20 "ValueEquation"
          Comment@20..21
            WHITESPACE@20..21 " "
        ModuleExportList@21..21
        WHERE@21..26 "where"
        Comment@26..28
          WHITESPACE@26..28 "\n\n"
      LAYOUT_START@28..28 ""
      ModuleImports@28..28
      ModuleStatements@28..78
        ValueEquation@28..78
          LOWER@28..29 "p"
          Comment@29..30
            WHITESPACE@29..30 " "
          EquationBinders@30..34
            BinderVariable@30..32
              LOWER@30..31 "a"
              Comment@31..32
                WHITESPACE@31..32 " "
            BinderVariable@32..34
              LOWER@32..33 "b"
              Comment@33..34
                WHITESPACE@33..34 " "
          EquationUnguarded@34..78
            EQUAL@34..35 "="
            Comment@35..36
              WHITESPACE@35..36 " "
            EquationWhere@36..78
              ExpressionTyped@36..46
                ExpressionOperatorChain@36..46
                  ExpressionVariable@36..39
                    LOWER@36..38 "a2"
                    Comment@38..39
                      WHITESPACE@38..39 " "
                  OPERATOR@39..40 "+"
                  Comment@40..41
                    WHITESPACE@40..41 " "
                  ExpressionVariable@41..46
                    LOWER@41..43 "b2"
                    Comment@43..46
                      WHITESPACE@43..46 "\n  "
              WHERE@46..51 "where"
              Comment@51..54
                WHITESPACE@51..54 "\n  "
              LetBindingStatements@54..78
                LAYOUT_START@54..54 ""
                LetBindingEquation@54..67
                  LOWER@54..56 "a2"
                  Comment@56..57
                    WHITESPACE@56..57 " "
                  EquationBinders@57..57
                  EquationUnguarded@57..67
                    EQUAL@57..58 "="
                    Comment@58..59
                      WHITESPACE@58..59 " "
                    EquationWhere@59..67
                      ExpressionTyped@59..67
                        ExpressionOperatorChain@59..67
                          ExpressionVariable@59..61
                            LOWER@59..60 "a"
                            Comment@60..61
                              WHITESPACE@60..61 " "
                          OPERATOR@61..62 "*"
                          Comment@62..63
                            WHITESPACE@62..63 " "
                          ExpressionVariable@63..67
                            LOWER@63..64 "a"
                            Comment@64..67
                              WHITESPACE@64..67 "\n  "
                LAYOUT_SEPARATOR@67..67 ""
                LetBindingEquation@67..78
                  LOWER@67..69 "b2"
                  Comment@69..70
                    WHITESPACE@69..70 " "
                  EquationBinders@70..70
                  EquationUnguarded@70..78
                    EQUAL@70..71 "="
                    Comment@71..72
                      WHITESPACE@71..72 " "
                    EquationWhere@72..78
                      ExpressionTyped@72..78
                        ExpressionOperatorChain@72..78
                          ExpressionVariable@72..74
                            LOWER@72..73 "b"
                            Comment@73..74
                              WHITESPACE@73..74 " "
                          OPERATOR@74..75 "*"
                          Comment@75..76
                            WHITESPACE@75..76 " "
                          ExpressionVariable@76..78
                            LOWER@76..77 "b"
                            Comment@77..78
                              WHITESPACE@77..78 "\n"
                LAYOUT_END@78..78 ""
      LAYOUT_END@78..78 ""
      END_OF_FILE@78..78 ""
    ,
    [],
)
